src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+;; You need to send a NAME_RENEWAL every so often to keep your name.
+;; You will pay the registration cost of your name to the namespace's designated burn address when you renew it.
+;; When a name expires, it enters a month-long "grace period" (5000 blocks). 
+;; It will stop resolving in the grace period, and all of the above operations will cease to be honored by the BNS consensus rules.
+;; You may, however, send a NAME_RENEWAL during this grace period to preserve your name.
+;; If your name is in a namespace where names do not expire, then you never need to use this transaction.
+(define-public (name-renewal (namespace (buff 19))
+                             (name (buff 16))
+                             (stx-to-burn uint)
+                             (new-owner (optional principal))
+                             (zonefile-hash (optional (buff 20))))
+  (let (
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (owner (unwrap!
+      (nft-get-owner? names (tuple (name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+    (name-props (unwrap!
+      (map-get? name-properties ((name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND)))) ;; The name must exist
+    ;; The namespace must be launched
+    (asserts!
+      (is-some (get launched-at namespace-props))
+      (err ERR_NAMESPACE_NOT_LAUNCHED))
+    ;; The namespace should require renewals
+    (asserts!
+      (> (get renewal-rule namespace-props) u0)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The sender must match the name's current owner
+    (asserts!
+      (is-eq owner contract-caller)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The name must have been registered (vs imported)
+    (unwrap!
+      (get registered-at name-props)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; If expired, the name must not be in the renewal grace period.
+    (if (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC))
+      (asserts!
+        (is-eq (unwrap! (is-name-in-grace-period? namespace name) (err ERR_PANIC)) true)
+        (err ERR_NAME_EXPIRED))
+      true)    
+    ;; The amount burnt must be equal to or greater than the cost of the namespace
+    (asserts!
+      (>= stx-to-burn (compute-name-price name (get price-function namespace-props)))
+      (err ERR_NAME_STX_BURNT_INSUFFICIENT))
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Transfer the name, if any new-owner
+    (if (is-none new-owner)
+      (ok true) 
+      (let ((owner-unwrapped (unwrap! new-owner (err ERR_PANIC))))
+        (let ((current-owned-name (map-get? owner-name ((owner owner-unwrapped)))))
+          (let ((can-new-owner-get-name (if (is-none current-owned-name)
```

Comment:
The new owner may also receive the name if it only owns revoked names.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/19a6ffc687c7118230d4249df93cae24cea3d536